text/0517-io-os-reform.md

```diff
@@ -708,7 +712,129 @@ throughout IO, we can go on to explore the modules in detail.
 ### `std::fs`
 [std::fs]: #stdfs
 
-> To be added in a follow-up PR.
+The `fs` module will provide most of the functionality it does today,
+but with a stronger cross-platform orientation.
+
+Note that all path-consuming functions will now take an
+`AsPath`-bounded parameter for ergonomic reasons (this will allow
+passing in Rust strings and literals directly, for example).
+
+#### Free functions
+[Free functions]: #free-functions
+
+**Files**:
+
+* `copy`. Take `AsPath` bound.
+* `rename`. Take `AsPath` bound.
+* `remove_file` (renamed from `unlink`). Take `AsPath` bound.
+
+* `file_attr` (renamed from `stat`). Take `AsPath` bound. Yield a new
+  struct, `FileAttr`, with no public fields, but `size`, `kind` and
+  `perm` accessors. The various `os::platform` modules will offer
+  extension methods on this structure.
+
+* `set_perm` (renamed from `chmod`). Take `AsPath` bound, and a
+  `FilePermissions` value. The `FilePermissions` type will be revamped
+  as a struct with private implementation; see below.
+
+**Directories**:
+
+* `make_dir` (renamed from `mkdir`). Take `AsPath` bound.
+* `make_dir_all` (renamed from `mkdir_recursive`). Take `AsPath` bound.
+* `read_dir` (renamed from `readdir`). Take `AsPath` bound. Yield a
+  newtypes iterator, which yields a new type `DirEntry` which has an
+  accessor for `Path`, but will eventually provide other information
+  as well (possibly via platform-specific extensions).
+* `remove_dir` (renamed from `rmdir`). Take `AsPath` bound.
+* `remove_dir_all` (renamed from `rmdir_recursive`). Take
+  `AsPath` bound.
```

Comment:
I feel like "create" and "delete" are used rather than "make" and "remove." I don’t think the names are bad as they are, but I think they would be better if they matched the terminology people commonly use.

I checked the file manager on Windows and Linux Mint and one on Android, and all of them had a menu item to "delete" a file/directory (OS X just has "Move to Trash" and "Empty Trash"). Even the Linux man page for `rmdir()` says it "deletes" a directory. Similarly, the `mkdir()` and `open()`/`creat()` man pages say that they can "create" a file/directory. Here’s a random page I found while looking for something slightly different: <http://apple.stackexchange.com/q/73876/66147>. There are multiple uses of "create" and "delete" (in some form) on the page.

I would rename:

- `remove_file` to `delete_file`
- `make_dir` to `create_dir`
- `make_dir_all` to `create_dir_all`
- `remove_dir` to `delete_dir`
- `remove_dir_all` to `delete_dir_all`